force-app/main/default/classes/Gifts.cls

```diff
@@ -159,6 +159,18 @@ public with sharing class Gifts {
         }
     }
 
+    public void failGiftsInAbortedOrFailedBatches(String failReason) {
```

Comment:
Can this be made into two reusable methods?  The method call could be: gifts.fail(gifts.getGiftsInProgress());